python/cudf/cudf/core/column/struct.py

```diff
@@ -23,10 +26,39 @@ class StructColumn(ColumnBase):
 
     Every column has n children, where n is
     the number of fields in the Struct Dtype.
-
     """
 
-    dtype: StructDtype
+    def __init__(
+        self,
+        data: None,
+        size: int,
+        dtype: StructDtype,
+        mask: Buffer | None = None,
+        offset: int = 0,
+        null_count: int | None = None,
+        children: tuple[ColumnBase, ...] = (),
+    ):
+        if data is not None:
+            raise ValueError("data must be None.")
+        dtype = self._validate_dtype_instance(dtype)
+        super().__init__(
+            data=data,
+            size=size,
+            dtype=dtype,
+            mask=mask,
+            offset=offset,
+            null_count=null_count,
+            children=children,
+        )
+
+    @staticmethod
+    def _validate_dtype_instance(dtype: StructDtype) -> StructDtype:
+        # IntervalDtype is a subclass of StructDtype, so compare types exactly
+        if type(dtype) is not StructDtype:
```

Comment:
It feels wrong to do this only for StructDtype because we know that other dtypes inherit from it, but to not do this in other cases. Should all validation functions like this be this strict, or are there actually cases where we want to allow for subclasses? Could we write a common implementation in the parent class that relies on a class member indicating the column's dtype? It seems like that's something we would want in general; every column type should know the dtype it's meant to hold.

Reply:
Yeah agreed it feels a little wrong to special case struct here. 

> Should all validation functions like this be this strict, or are there actually cases where we want to allow for subclasses? 

If we were fully strict like StructColumn, we could be potentially prevent user-subclassed dtypes from public constructors which would eventually be stored here on the column (not sure what's cudf's current promise on subclass compatibility).

> Could we write a common implementation in the parent class that relies on a class member indicating the column's dtype?

Yeah generally agreed that this would be good to have. I can explore this in a follow up.

Reply:
Subclassing dtypes and column types is something we don't support well right now, but IMO we should. Providing proper public entry points there would allow libraries like cuspatial to extend the library without needing to reach into our internals the way that they do now. I discuss this more in #12495. We should think carefully about this when we're rearchitecting our internals around pylibcudf.